### ISSUE TITLE
0.50 / 0.50.1 breaks the stickler mirror command

### ISSUE THREAD

**Issue by Iristyle** [excon/excon#579]
Stickler, at version `2.4.2` has remained unchanged since 3/13/2015.  However, installing a fresh copy of stickler will pull in `excon ~> 0.37`, which currently pulls `excon` version `0.50.1`.  Attempting to use stickler to mirror a gem will fail with this version of the `excon` gem:

```
stickler mirror --server http://rubygems.delivery.puppetlabs.net ffi --gem-version 1.9.13 --platform=x86-mingw32

Asking http://rubygems.delivery.puppetlabs.net/ to mirror ffi-1.9.13-x86-mingw32 from rubygems.org : /usr/local/opt/rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/excon-0.50.1/lib/excon/middlewares/expects.rb:10:in `response_call'
/usr/local/opt/rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/excon-0.50.1/lib/excon/middlewares/response_parser.rb:8:in `response_call'
/usr/local/opt/rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/excon-0.50.1/lib/excon/connection.rb:389:in `response'
/usr/local/opt/rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/excon-0.50.1/lib/excon/connection.rb:253:in `request'
/usr/local/opt/rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/stickler-2.4.2/lib/stickler/repository/remote.rb:296:in `resource_request'
/usr/local/opt/rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/stickler-2.4.2/lib/stickler/repository/remote.rb:268:in `remote_uri_exist?'
/usr/local/opt/rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/stickler-2.4.2/lib/stickler/repository/remote.rb:264:in `remote_gem_file_exist?'
/usr/local/opt/rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/stickler-2.4.2/lib/stickler/repository/remote_mirror.rb:15:in `mirror'
/usr/local/opt/rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/stickler-2.4.2/lib/stickler/client/mirror.rb:69:in `mirror_one_spec'
/usr/local/opt/rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/stickler-2.4.2/lib/stickler/client/mirror.rb:86:in `block in run'
/usr/local/opt/rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/stickler-2.4.2/lib/stickler/client/mirror.rb:85:in `each'
/usr/local/opt/rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/stickler-2.4.2/lib/stickler/client/mirror.rb:85:in `run'
/usr/local/opt/rbenv/versions/1.9.3-p551/lib/ruby/gems/1.9.1/gems/stickler-2.4.2/bin/stickler:55:in `<top (required)>'
/usr/local/opt/rbenv/versions/1.9.3-p551/bin/stickler:22:in `load'
/usr/local/opt/rbenv/versions/1.9.3-p551/bin/stickler:22:in `<main>'
ERROR -> Expected([200]) <=> Actual(404 Not Found)
```

If I remove the `excon` version that was pulled in by `gem install`, and then manually `gem install excon --version 0.49.0`, I can successfully mirror again:

```
stickler mirror ffi --gem-version 1.9.13 --platform=x86-mingw32 --upstream http://rubygems.org --server http://rubygems.delivery.puppetlabs.net

Asking http://rubygems.delivery.puppetlabs.net/ to mirror ffi-1.9.13-x86-mingw32 from rubygems.org : OK -> http://rubygems.delivery.puppetlabs.net/gems/ffi-1.9.13-x86-mingw32.gem
```

I also tried to use the `0.50` version of excon, which resulted in the same problem. Therefore the cause of the failure was introduced somewhere in here: https://github.com/excon/excon/compare/v0.49.0...v0.50.0

**Comment by geemus** [comment]
Thanks for the detailed report. I'm not particularly familiar with stickler, so not sure off hand where to go from there. Could you try running it with `EXCON_DEBUG=true` set to get more detailed output? Otherwise I could venture rough guesses, but it would be better to see for sure. If you actually could share output from success and failure cases that would be even more helpful to find the difference so that I can work backwards from there. Thanks!

**Comment by starbelly** [comment]
@Iristyle Replicated, trying to figure out why...

**Comment by starbelly** [comment]
@Iristyle Right now this seems like a stickler problem. If I follow the docs, I can replicate this, but if I go to localhost:6789/... in my browser, I get a 404 there too. stickler is not attempting to catch that, so not sure what the problem is... 

Here is excon debug: 

```
Asking http://localhost:6789/ to mirror ffi-1.9.13-x86-mingw32 from rubygems.org : excon.request
  :chunk_size          => 1048576
  :ciphers             => "HIGH:!SSLv2:!aNULL:!eNULL:!3DES"
  :connect_timeout     => 60
  :debug_request       => true
  :debug_response      => true
  :expects             => [
    200
  ]
  :headers             => {
    "Host"       => "localhost:6789"
    "User-Agent" => "Stickler Client v2.4.2"
  }
  :host                => "localhost"
  :hostname            => "localhost"
  :idempotent          => false
  :instrumentor        => Excon::StandardInstrumentor
  :instrumentor_name   => "excon"
  :method              => :head
  :middlewares         => [
    Excon::Middleware::ResponseParser
    Excon::Middleware::Expects
    Excon::Middleware::Idempotent
    Excon::Middleware::Instrumentor
    Excon::Middleware::Mock
  ]
  :mock                => false
  :nonblock            => true
  :omit_default_port   => false
  :path                => "/gems/ffi-1.9.13-x86-mingw32.gem"
  :persistent          => false
  :port                => 6789
  :query               => nil
  :read_timeout        => 60
  :retries_remaining   => 4
  :retry_limit         => 4
  :scheme              => "http"
  :ssl_uri_schemes     => [
    "https"
  ]
  :ssl_verify_peer     => true
  :stubs               => :global
  :tcp_nodelay         => false
  :thread_safe_sockets => true
  :uri_parser          => URI
  :versions            => "excon/0.50.1 (x86_64-darwin15) ruby/2.3.1"
  :write_timeout       => 60
excon.error
  :error => #<Excon::Error::NotFound: Expected([200]) <=> Actual(404 Not Found)
excon.error.request
  :chunk_size          => 1048576
  :ciphers             => "HIGH:!SSLv2:!aNULL:!eNULL:!3DES"
  :connect_timeout     => 60
  :debug_request       => true
  :debug_response      => true
  :expects             => [
    200
  ]
  :headers             => {
    "Content-Length" => 0
    "Host"           => "localhost:6789"
    "User-Agent"     => "Stickler Client v2.4.2"
  }
  :host                => "localhost"
  :hostname            => "localhost"
  :idempotent          => false
  :instrumentor        => Excon::StandardInstrumentor
  :instrumentor_name   => "excon"
  :method              => :head
  :middlewares         => [
    Excon::Middleware::ResponseParser
    Excon::Middleware::Expects
    Excon::Middleware::Idempotent
    Excon::Middleware::Instrumentor
    Excon::Middleware::Mock
  ]
  :mock                => false
  :nonblock            => true
  :omit_default_port   => false
  :path                => "/gems/ffi-1.9.13-x86-mingw32.gem"
  :persistent          => false
  :port                => 6789
  :query               => nil
  :read_timeout        => 60
  :retries_remaining   => 4
  :retry_limit         => 4
  :scheme              => "http"
  :ssl_uri_schemes     => [
    "https"
  ]
  :ssl_verify_peer     => true
  :stubs               => :global
  :tcp_nodelay         => false
  :thread_safe_sockets => true
  :uri_parser          => URI
  :versions            => "excon/0.50.1 (x86_64-darwin15) ruby/2.3.1"
  :write_timeout       => 60
excon.error.response
  :body          => ""
  :cookies       => [
  ]
  :headers       => {
    "Cache-Control"          => "no-cache"
    "Connection"             => "Keep-Alive"
    "Content-Length"         => "1712"
    "Content-Type"           => "text/html"
    "Date"                   => "Thu, 07 Jul 2016 13:45:25 -0500"
    "Server"                 => "WEBrick/1.3.1 (Ruby/2.3.1/2016-04-26)"
    "X-Content-Type-Options" => "nosniff"
    "X-Frame-Options"        => "SAMEORIGIN"
    "X-Xss-Protection"       => "1; mode=block"
  }
  :host          => "localhost"
  :local_address => "127.0.0.1"
  :local_port    => 53276
  :path          => "/gems/ffi-1.9.13-x86-mingw32.gem"
  :port          => 6789
  :reason_phrase => "Not Found "
  :remote_ip     => "::1"
  :status        => 404
  :status_line   => "HTTP/1.1 404 Not Found \r\n"
```

 Still looking...

**Comment by geemus** [comment]
My guess would be that port is included in one request, but not the other. Though I don't think there have been any changes that I can remember that would cause that particular issue. It does seem like an issue with stickler, in that it doesn't handle a 404 very nicely, but I would expect that to fail in the same way in both places. That said, I suppose it is possible that our refactoring of error classes is at fault (such that stickler tries to rescue the error, but misses because the classes changed)?

**Comment by starbelly** [comment]
@geemus It has something to do with attempting to catch Excon::Errors::Error...  Specifically, the method `remote_uri_exists?' tries to catch Excon::Errors::Error and return false.  Trying to figure out why now.

**Comment by starbelly** [comment]
Confirmed, Excon::Errors::Error isn't acting as catch all, which it should.. Putting in a regression test for this as soon as I figure exactly why. 

Ancestry for Excon::Errors::Error looks like so:

[Excon::Errors::Error, Excon::Error, StandardError, Exception, Object, Kernel, BasicObject]

Gonna go grab a bite to eat then come back to this.

**Comment by geemus** [comment]
Strange, not sure off hand why that would be (it seemed like it should be working to me). Enjoy the food and let me know if I can be of further assistance.

**Comment by starbelly** [comment]
@geemus https://github.com/excon/excon/pull/581

@Iristyle This should fix the issue.

**Comment by geemus** [comment]
Should now be fixed and released in 0.51.0. Thanks!
